metricbeat/module/vsphere/datastore_usage/datastore_usage.go

```diff
@@ -0,0 +1,130 @@
+package datastore_usage
+
+import (
+	"context"
+	"errors"
+	"net/url"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/metricbeat/mb"
+
+	"github.com/vmware/govmomi"
+	"github.com/vmware/govmomi/find"
+	"github.com/vmware/govmomi/property"
+	"github.com/vmware/govmomi/vim25"
+	"github.com/vmware/govmomi/vim25/mo"
+	"github.com/vmware/govmomi/vim25/types"
+)
+
+func init() {
+	if err := mb.Registry.AddMetricSet("vsphere", "datastore_usage", New); err != nil {
+		panic(err)
+	}
+}
+
+type MetricSet struct {
+	mb.BaseMetricSet
+	Client *vim25.Client
+}
+
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+
+	config := struct {
+		Username string `config:"username"`
+		Password string `config:"password"`
+		Insecure bool   `config:"insecure"`
+	}{}
+
+	if err := base.Module().UnpackConfig(&config); err != nil {
+		return nil, err
+	}
+
+	u, err := url.Parse(base.HostData().URI)
+	if err != nil {
+		return nil, err
+	}
+
+	u.User = url.UserPassword(config.Username, config.Password)
+
+	c, err := govmomi.NewClient(context.TODO(), u, config.Insecure)
+	if err != nil {
+		return nil, err
+	}
+
+	return &MetricSet{
+		BaseMetricSet: base,
+		Client:        c.Client,
+	}, nil
+}
+
+func (m *MetricSet) Fetch() ([]common.MapStr, error) {
+
+	f := find.NewFinder(m.Client, true)
+	if f == nil {
+		return nil, errors.New("Finder undefined for vsphere.")
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// Get all datacenters
+	dcs, err := f.DatacenterList(ctx, "*")
+	if err != nil {
+		return nil, err
+	}
+
+	events := []common.MapStr{}
+
+	for _, dc := range dcs {
+
+		f.SetDatacenter(dc)
+
+		dss, err := f.DatastoreList(ctx, "*")
+		if err != nil {
+			return nil, err
+		}
+
+		pc := property.DefaultCollector(m.Client)
+
+		// Convert datastores into list of references
+		var refs []types.ManagedObjectReference
+		for _, ds := range dss {
+			refs = append(refs, ds.Reference())
+		}
+
+		// Retrieve summary property
+		var dst []mo.Datastore
+		err = pc.Retrieve(ctx, refs, []string{"summary"}, &dst)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, ds := range dst {
+
+			usedSpacePercent := 100 * (ds.Summary.Capacity - ds.Summary.FreeSpace) / ds.Summary.Capacity
```

Comment:
We store percentages as "raw" values means the `100 *` is not needed. Kibana will automatically make the transformation if it is mentioned in the index pattern (see comment above).

In general we try not to have data in the metricset which can also be directly calculated on the ES / Kibana side which I think is the case with the pct value here. I see the reason why you added it as it is probably one of the most important values to watch for. @tsg any thoughts?